### PR TITLE
Software bill of materials API

### LINKED ISSUE #4374: Software bill of materials API

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #4374

### Endpoint

* `GET` `/api/v1/applications/:applicationId/bom`

### Added tests
`test/unit/forge/db/models/Application_spec.js` (new file)
```
  Application model
    Relations
      ✔ should delete Application when team is deleted
    instance methods
      getChildren
        ✔ should return an child instances, devices and instance devices (50ms)
        ✔ should return an child instances, devices and instance devices with dependencies (40ms)
```
`test/unit/forge/ee/routes/api/application_spec.js` (new file)
```
  Application API
    BOM
      ✔ Owner can get Application BOM (122ms)
      ✔ Non Owner can not get BOM (403) (84ms)
      ✔ Non Member can not get BOM (404) (80ms)
```

### Notes:

* there was a small amount or refactoring in `forge/routes/api/deviceLive.js` to move versioned modules that get written into snapshots into the device instance model so that the BOM can utelise them without duplicating the logic
* the BOM is currently a flat structure as was proposed in the issue comments, but for future thinking, I have proposed a small tweak that makes it a nested structure (below). I think makes more sense from the get go (avoids later refactoring)
* In order to permit visible organisation of devices and for other reasons (like generating links based on instance owned etc) I included the `ownerType` and `ownerId`. Note: these become redundant if the format is changed to "nested" as per my notes below.

#### Instances

* 🗒️ The "installed" version of Node-RED is not reported or stored in the DB. It is assumed from the project stack setting.
* 🗒️ The "semver" version of Node-RED is also taken from the project stack however if the stack has a `replacedBy` value, the semver version is updated to that.
* 🗒️ "Installed" nodes versions are taken from ProjectSettings table
* 🗒️ "semver" versions are taken from the palette settings for an instance. 
* 🔥 Some nodes (e.g. `flowfuse/nr-project-nodes`, `@flowfuse/nr-file-nodes`, `@flowfuse/nr-assistant`) are not known to core but might be out of date.
* 🔥 modules installed to a function node end up in palette modules as version '*' but installed version is never reported

#### Devices (application)

* 🔥 flowforge-library-plugin is the team lib plugin. This is an embedded module in the device agent. It shows up in Node-RED palette manager but it exists nowhere in NPM/FF Core and is not reported back to core.
* 🔥 "installed" versions (including "Node-RED" itself) are (mostly) assumed. For example, if the DB device has an "active snapshot", this code takes the modules as being installed. However, the device could have been updated at any time without a new snapshot taken or marked as active.  In short, devices do NOT report an inventory of actual versions.
* 🔥 The "semver" Node-RED version is not a single fixed data point in the database. It could be from the target snapshot. If there is no target it could be from the starter snapshot (i.e. "latest"). It could be "3.0.2" if the launcher version is not populated. All of this makes things tricky.
* 🔥 modules installed in a function node do not end up available to core unless a snapshot is taken and made active/target.

#### Devices (instance)
* 🗒️ A mix of the above in that it "should" get modules specified by the instance however the target snapshot may be older.
* 🔥 modules installed in a function node do not end up in the instance.

#### Improvement notes
* 💡 We need instances and devices to report exactly what they are told to install and what they running for this data to be reliable.
  * A potential source of data is the node-red system diagnostics API. It includes module versions (including node-red version). It also includes the Node-JS version that NR is running on.

#### Flat format example (current implementation inline with issue proposal)
```json
{
  "id": "QJRYOX1ej8",
  "name": "Application 1",
  "children": [
    {
      "id": "3b39f2c4-eb4d-43dd-99b0-a47ca106cd40",
      "name": "system-instance-1",
      "type": "instance",
      "dependencies": [
        {
          "name": "node-red",
          "version": {
            "semver": "3.0.2",
            "installed": "3.0.2"
          }
        },
        {
          "name": "node-red-contrib-buffer-parser",
          "version": {
            "semver": "latest",
            "installed": "3.2.2"
          }
        },
        {
          "name": "node-red-contrib-omron-fins",
          "version": {
            "semver": "latest",
            "installed": "0.5.0"
          }
        },
        {
          "name": "date-fns",
          "version": {
            "semver": "*",
            "installed": null
          }
        }
      ]
    },
    {
      "id": "bb362abb-6cc3-49af-94cd-d48c27a68509",
      "name": "dev-instance-no-2",
      "type": "instance",
      "dependencies": [...]
    },
    {
      "id": "36LO0L3Odb",
      "name": "Dev 2, Device 1",
      "type": "device",
      "ownerType": "instance",
      "ownerId": "bb362abb-6cc3-49af-94cd-d48c27a68509",
      "dependencies": [...]
    },
    {
      "id": "PRbDwkgDVW",
      "name": "instance-1-device-1",
      "type": "device",
      "ownerType": "instance",
      "ownerId": "3b39f2c4-eb4d-43dd-99b0-a47ca106cd40",
      "dependencies": [...]
    },
    {
      "id": "DScDwlpLFP",
      "name": "instance-1-device-2",
      "type": "device",
      "ownerType": "instance",
      "ownerId": "3b39f2c4-eb4d-43dd-99b0-a47ca106cd40",
      "dependencies": [...]
    },
    {
      "id": "jKJ1KBnDPq",
      "name": "App Device 1",
      "type": "device",
      "ownerType": "application",
      "ownerId": "QJRYOX1ej8",
      "dependencies": [...]
    },
    {
      "id": "RgeDWe29az",
      "name": "App Device 2",
      "type": "device",
      "ownerType": "application",
      "ownerId": "QJRYOX1ej8",
      "dependencies": [...]
    }
  ]
}
```

## Related Issue(s)

owner #4374
parent #4362

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
Upon completing the work for this and noting the future requirement for team level BOM, I took it one step further to see if it still "fits"

It is my opinion we would have more forward flexibility if we nest the data from the get-go. A future structure could make use of an "ou" (organisational unit) while maintaining the inner structure as it would be today (in the nested format that is) 

#### Nested format example (alt proposal - similar to issue proposal but devices are nested when owned by an instance)

```json
{
  "id": "QJRYOX1ej8",
  "name": "Application 1",
  "type": "application",
  "children": [
    {
      "id": "3b39f2c4-eb4d-43dd-99b0-a47ca106cd40",
      "name": "system-instance-1",
      "type": "instance",
      "dependencies": [...],
      "children": [
        {
          "id": "PRbDwkgDVW",
          "name": "instance-1-device-1",
          "type": "device",
          "dependencies": [...]
        },
        ... device
        ... device
        ... device
      ]
    },
    ,
    {
      "id": "bb362abb-6cc3-49af-94cd-d48c27a68509",
      "name": "dev-instance-no-2",
      "type": "instance",
      "dependencies": [...],
      "children": [
        {
        ... device
        ... device
        }
      ]
    },
    {
      "id": "jKJ1KBnDPq",
      "name": "App Device 1",
      "type": "device",
      "dependencies": [...]
    },
    {
      "id": "RgeDWe29az",
      "name": "App Device 2",
      "type": "device",
      "dependencies": [...]
    }
  ]
}
```

#### Possible Team BOM example if we DO change this up to a nested format
```json
{
  "id": "Fr77792c4",
  "name": "Team 1",
  "type": "team",
  "children": [
    {
      "id": "QJRYOX1ej8",
      "name": "Application 1",
      "type": "application",
      "children": [
        ... instance
        ... instance
          ... device
          ... device
        ... instance
        ... device
        ... device
      ]
    },
    {
      "id": "FRPOPX4ab9",
      "name": "Application 2",
      "type": "application",
      "children": [
        ... instance
          ... device
          ... device
        ... device
      ]
    }
  ]
}
```

#### Further ahead thinking - nested format still holds true

##### site level

```json
{
  "id": "gR12392dd",
  "name": "Site 1",
  "type": "ou",
  "children": [
    ... application
      ... instance
      ... instance
        ... device
        ... device
      ... instance
      ... device
      ... device
    ... application
      ... instance
        ... device
    ... device
  ]
}
```

##### corporate level

```json
{
  "id": "eR12392dd",
  "name": "Enterprise 1",
  "type": "ou",
  "children": [
    {
      "id": "gR12392dd",
      "name": "Site 1",
      "type": "ou",
      "children": [
        ...team
          ... application
            ... device
            ... device
          ... application
            ... instance
              ... device
              ... device
            ... instance
            ... device
        ...team
          ... application
            ... instance
              ... device
            ... instance
            ... device
          ... application
          ... application
      ]
    },
    {
      "id": "gR12392dd",
      "name": "Site 2",
      "type": "ou",
      "children": [
        ...team
          ...application
            ...instance
            ...instance
          ...application
            ... instance
              ... device
              ... device
            ... instance
            ... device
            ... device
      ]
    }
  ]
}
```

**Comment by Steve-Mcl** [comment]
@knolleary I have requested an early review (sans tests) in the hope you can give this a once over to steer it if necessary before I pick it back up and write tests etc. Ta.

**Comment by knolleary** [comment]
@Steve-Mcl I've been heads down on the tasks I already had to squeeze in today - didn't seen your notification until now. I will see if I can take a look at this before I head off - but I'm pretty much out of time.

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl I've been heads down on the tasks I already had to squeeze in today - didn't seen your notification until now. I will see if I can take a look at this before I head off - but I'm pretty much out of time.

Dont worry about it Nick. Enjoy your downtime.
I'll make some executive decisions and crack on with it. If some rejigging is required, so be it.

File: frontend/src/pages/admin/TeamTypes/dialogs/TeamTypeEditDialog.vue [hunk 1]
```diff
@@ -101,7 +101,9 @@
                     <FormRow v-model="input.properties.features.editorLimits" type="checkbox">API/Debug Length Limits</FormRow>
                     <FormRow v-model="input.properties.features.customHostnames" type="checkbox">Custom Hostnames</FormRow>
                     <FormRow v-model="input.properties.features.staticAssets" type="checkbox">Static Assets</FormRow>
-                    <span /> <!-- to make the grid work nicely, only needed if there is an odd number of checkbox features above -->
+                  <!-- to make the grid work nicely, only needed if there is an odd number of checkbox features above-->
+                  <!-- <span />-->
```
Comment (knolleary) [inline frontend/src/pages/admin/TeamTypes/dialogs/TeamTypeEditDialog.vue:105]:
```suggestion
                    <!-- to make the grid work nicely, only needed if there is an odd number of checkbox features above-->
                    <!-- <span />-->
```

Fix linting warning

**Review by knolleary — APPROVED** [review]
Hey @Steve-Mcl - finally picking this up - sorry for the delay.

Based on your comment:

> the BOM is currently a flat structure as was proposed in the issue comments, but for future thinking, I have proposed a small tweak that makes it a nested structure (below). I think makes more sense from the get go (avoids later refactoring)

I suggest we keep the implemented flat structure for the Application level; we can nest Applications when providing a Team level BOM.

One minor linting proposed change

**Comment by knolleary** [comment]
Failing backend tests:
```
  1) Application API
       BOM
         Owner can get Application BOM:

      AssertionError: expected 404 to be 200
      + expected - actual

      -404
      +200
      
      at Assertion.fail (node_modules/should/cjs/should.js:275:17)
      at Assertion.value (node_modules/should/cjs/should.js:356:19)
      at Context.<anonymous> (test/unit/forge/ee/routes/api/application_spec.js:110:40)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Review by knolleary — APPROVED** [review]

**Comment by knolleary** [comment]
Tests all failing

**Comment by cstns** [comment]
https://github.com/FlowFuse/flowfuse/pull/4466 removed one of the pictograms used in the bom UI.

**Comment by cstns** [comment]
not sure why unit tests were failing though, ran everything successfully locally

**Comment by Steve-Mcl** [comment]
@cstns regarding no Node-RED version in software BOM on staging...

I have spoke with @hardillb and confirmed the stacks table version is not populated for Docker or K8s.

This is another shortfall of not having instances and devices report up (i.e. single source of truth)

How to processed needs to be discussed. Immediate options include:
* halt enabling the feature until we have NR versions populating a table (or we query each launcher (note only running launchers will respond))
* handle missing NR versions 
   * opt1: dont report on Node-RED version
   * opt2: place them in a not available row
   * opt3: aysnc request them from the launcher when the page is viewed 
   * other?

**Pull request merged**